Inline assembly (#2850)
rust-lang/rfcs: text/0000-inline-asm.md

- Feature Name: `asm`
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This RFC specifies a new syntax for inline assembly which is suitable for eventual stabilization.

> The initial implementation of this feature will focus on the ARM, x86 and RISC-V architectures. Support for more architectures will be added based on user demand.

Author: Yes, register definitions are basically the only thing needed to add support for a new architecture. This should be fairly straightforward once the basic infrastructure for inline asm is implemented.

> The initial implementation of this feature will focus on the ARM, x86 and RISC-V architectures. Support for more architectures will be added based on user demand.

Reviewer: >register definitions are basically the only thing needed to add support for a new architecture. This should be fairly straightforward

Unless it's something alien, like e.g. [Intel GPU ISA](https://software.intel.com/en-us/articles/introduction-to-gen-assembly) :smile:

> The initial implementation of this feature will focus on the ARM, x86 and RISC-V architectures. Support for more architectures will be added based on user demand.

Reviewer 2: In practice the backend must have the inline assembly support for said target too. Not all of them do.

> The existing `asm!` macro will be renamed to `llvm_asm!` to provide an easy way to maintain backwards-compatibility with existing code using inline asm. However `llvm_asm!` is not intended to ever be stabilized.

Reviewer 3: Will there be some transition period where both `asm!` and `llvm_asm!` work like the current `asm!` and some lint warns against using `asm!`? I know `asm!` is unstable and all, but quite a few crates use the current `asm!` and changing the semantics of it without some heads-up could be a bit frustrating? Doesn't even need to be very long, 6 weeks or so should probably do.

Author: Yes that is the intent, see #2843 for details on `llvm_asm!`.

Reviewer 3: Ah perfect! A link that that RFC could be included in this RFC then :)

# Motivation
[motivation]: #motivation

> In systems programming some tasks require dropping down to the assembly level. The primary reasons are for performance, precise timing, and low level hardware access. Using inline assembly for this is sometimes convenient, and sometimes necessary to avoid function call overhead.

Reviewer 4: shout out to https://crates.io/crates/x86

> In systems programming some tasks require dropping down to the assembly level. The primary reasons are for performance, precise timing, and low level hardware access. Using inline assembly for this is sometimes convenient, and sometimes necessary to avoid function call overhead.

Author: Also https://github.com/bjorn3/inline_asm_catalogue/

> In systems programming some tasks require dropping down to the assembly level. The primary reasons are for performance, precise timing, and low level hardware access. Using inline assembly for this is sometimes convenient, and sometimes necessary to avoid function call overhead.

Reviewer 5: * [`cortex-m`](https://github.com/rust-embedded/cortex-m), [`riscv`](https://github.com/rust-embedded/riscv) and [`msp430`](https://github.com/rust-embedded/msp430) use a lot of assembly to provide single-instruction intrinsics and access to processor registers
* [`cortex-m-rt`](https://github.com/rust-embedded/cortex-m-rt/) uses a small assembly shim to pass an argument to the `HardFault` handler
* [`riscv-rt`](https://github.com/rust-embedded/riscv-rt) uses assembly for processor initialization code
* ... probably others I forgot about

> In systems programming some tasks require dropping down to the assembly level. The primary reasons are for performance, precise timing, and low level hardware access. Using inline assembly for this is sometimes convenient, and sometimes necessary to avoid function call overhead.

Reviewer 6: This seems to contradict a later paragraph about [compiler support](https://github.com/Amanieu/rfcs/blob/inline-asm/text/0000-inline-asm.md#difficulty-of-support). The example below it advocates for moving assembly into an extern "C" function interface which would incurr the function call overhead and lose the control valued in the motivation.

> However it is possible to implement support for inline assembly without support from the compiler backend by using an external assembler instead.

Which is then no longer inline assembly but rather an interesting way to interact with the linker from within Rust code. A different concept, imho, although also an interesting one.

> In systems programming some tasks require dropping down to the assembly level. The primary reasons are for performance, precise timing, and low level hardware access. Using inline assembly for this is sometimes convenient, and sometimes necessary to avoid function call overhead.

Reviewer 7: If I understand correctly, this is one of those "if we have to live like this we will, but we don't want to" sort of fallbacks.

But that should absolutely be cleared up more in the wording.

> In systems programming some tasks require dropping down to the assembly level. The primary reasons are for performance, precise timing, and low level hardware access. Using inline assembly for this is sometimes convenient, and sometimes necessary to avoid function call overhead.

Reviewer 8: I needed the `divq` x86_64 instruction to get the best u128 division performance, see my [compiler-rt PR](https://github.com/rust-lang/compiler-builtins/pull/332)

Generally this should not be necessary, but might be where the required performance or timing
cannot be otherwise achieved. Accessing low level hardware primitives, e.g. in kernel code, may also demand this functionality.

> Note: the examples here are given in x86/x86-64 assembly, but ARM, AArch64 and RISC-V are also supported.

## Basic usage

Let us start with the simplest possible example:

```rust
unsafe {
    asm!("nop");
}
```

This will insert a NOP (no operation) instruction into the assembly generated by the compiler.
Note that all `asm!` invocations have to be inside an `unsafe` block, as they could insert
arbitrary instructions and break various invariants. The instructions to be inserted are listed
in the first argument of the `asm!` macro as a string literal.

## Inputs and outputs

Now inserting an instruction that does nothing is rather boring. Let us do something that
actually acts on data:

```rust
let x: u32;
unsafe {
    asm!("mov {}, 5", out(reg) x);
}
assert_eq!(x, 5);
```

This will write the value `5` into the `u32` variable `x`.
You can see that the string literal we use to specify instructions is actually a template string.

> It is governed by the same rules as Rust [format strings][format-syntax].

Author: This is actually addressed in the reference-level explanation, which explicitly says that implicit named arguments are not supported. Adding support for a short-hand notation is listed as a future possibility, but we felt that it was not worth blocking the RFC on.

The arguments that are inserted into the template however look a bit different then you may
be familiar with. First we need to specify if the variable is an input or an output of the
inline assembly. In this case it is an output. We declared this by writing `out`.
We also need to specify in what kind of register the assembly expects the variable.
In this case we put it in an arbitrary general purpose register by specifying `reg`.
The compiler will choose an appropriate register to insert into
the template and will read the variable from there after the inline assembly finishes executing.

Let see another example that also uses an input:

```rust
let i: u32 = 3;
let o: u32;
unsafe {
    asm!("
        mov {0}, {1}
        add {0}, {number}
    ", out(reg) o, in(reg) i, number = imm 5);
}

> assert_eq!(i, 8);

Reviewer 9: Was this intended to test `o`?

```suggestion
assert_eq!(o, 8);
```

and then adding `5` to it.

The example shows a few things:

First we can see that inputs are declared by writing `in` instead of `out`.

Second one of our operands has a type we haven't seen yet, `imm`.
This tells the compiler to expand this argument to an immediate inside the assembly template.
This is only possible for constants and literals.

Third we can see that we can specify an argument number, or name as in any format string.
For inline assembly templates this is particularly useful as arguments are often used more than once.
For more complex inline assembly using this facility is generally recommended, as it improves
readability, and allows reordering instructions without changing the argument order.

We can further refine the above example to avoid the `mov` instruction:

```rust
let mut x: u32 = 3;
unsafe {
    asm!("add {0}, {number}", inout(reg) x, number = imm 5);
}
assert_eq!(x, 8);
```

We can see that `inout` is used to specify an argument that is both input and output.
This is different from specifying an input and output separately in that it is guaranteed to assign both to the same register.

> It is also possible to specify different variables for the input and output parts of an `inout` operand:

Author: One of the main uses for this is to indicate an input register which is clobbered. This is represented as `inout(reg) some_val => _`.

> It is also possible to specify different variables for the input and output parts of an `inout` operand:

Author: Currently only `Copy` types are supported as asm operands. But otherwise the input part is essentially treated the same way as a function argument.

> It is also possible to specify different variables for the input and output parts of an `inout` operand:

Reviewer 10: @bstrie Even if you could do it with a `mov` instruction, telling the compiler where you left the variable's value allows the compiler to just treat that as the new location of the variable, which means the resulting assembly won't need a `mov` at all.

```rust
let x: u32 = 3;
let y: u32;
unsafe {
    asm!("add {0}, {number}", inout(reg) x => y, number = imm 5);
}
assert_eq!(y, 8);
```

## Late output operands

The Rust compiler is conservative with its allocation of operands. It is assumed that an `out`
can be written at any time, and can therefore not share its location with any other argument.
However, to guarantee optimal performance it is important to use as few registers as possible,
so they won't have to be saved and reloaded around the inline assembly block.
To achieve this Rust provides a `lateout` specifier. This can be used on any output that is

> guaranteed to be written only after all inputs have been consumed.

Author: It is enforced by the user. In hindsight the word `guarantee` might be misleading here.

There is also a `inlateout` variant of this specifier.

Here is an example where `inlateout` *cannot* be used:

```rust
let mut a = 4;
let b = 4;
let c = 4;
unsafe {
    asm!("
        add {0}, {1}
        add {0}, {2}
    ", inout(reg) a, in(reg) b, in(reg) c);
}
assert_eq!(a, 12);
```

Here the compiler is free to allocate the same register for inputs `b` and `c` since it knows they have the same value. However it must allocate a separate register for `a` since it uses `inout` and not `inlateout`.

However the following example can use `inlateout` since the output is only modified after all input registers have been read:

```rust
let mut a = 4;
let b = 4;
unsafe {
    asm!("add {0}, {1}", inlateout(reg) a, in(reg) b);
}
assert_eq!(a, 8);
```

As you can see, this assembly fragment will still work correctly if `a` and `b` are assigned to the same register.

## Explicit register operands

Some instructions require that the operands be in a specific register.
Therefore, Rust inline assembly provides some more specific constraint specifiers.

> While `reg` is generally available on any architecture, these are highly architecture specific. E.g. for x86 the general purpose registers `eax`, `ebx`, `ecx`, `edx`, `ebp`, `esi`, and `edi`

Author: One possibility that was mentioned in discussions is WASM, which only supports a single type of operand: local variable indices.

> While `reg` is generally available on any architecture, these are highly architecture specific. E.g. for x86 the general purpose registers `eax`, `ebx`, `ecx`, `edx`, `ebp`, `esi`, and `edi`

Reviewer 11: LLVM treats Wasm's local variable indices as if they were registers for inline asm, using the "r" constraint. So `reg` would only not be available on Wasm if we explicitly changed that (e.g. we change it to `local` to make it clear that it's the index and not the operand stack).

among others can be addressed by their name.

```rust
unsafe {
    asm!("out 0x64, {}", in("eax") cmd);
}
```

In this example we call the `out` instruction to output the content of the `cmd` variable
to port `0x64`. Since the `out` instruction only accepts `eax` (and its sub registers) as operand

> we had to use the `eax` constraint specifier.

Reviewer 12: One of the problems I have with the current `asm!()` macro is that it does not at all feel like the error reporting story is on the same level of quality as the rest of `rustc`. See e.g. rust-lang/rust#15402

If I accidentally use `reg` here instead of `"eax"`, will I get poor error messages from deep inside the belly of LLVM?

Reviewer 13: Worth noting that your specific example won't be possible anymore, since as noted in the RFC, the input/output parameters will only accept raw numeric types (and pointers where it makes sense). Your wrapper struct would get rejected by the type checker.

There are other types of weird "low-level" llvm errors that can get reported when the feature is misused though. Notably, I remember getting really weird aborts when misusing constraints. Rustc will probably need to sanitize them before passing it to LLVM, making sure they make sense. Some things on the top of my head:

- Using the same named register twice as an input parameter (it's UB, LLVM accepts it and will either abort or does something weird)
- Using the same named register as both a clobber and an input or an output (UB, LLVM generates garbage, sometimes aborts)

Author: The specification in this RFC and the validation it requires should eliminate the possibility of any LLVM internal error. For example, the RFC explicitly says this:

> It is a compile-time error to use the same explicit register two input operand or two output operands.

Reviewer 12: In this case, the `out` instruction must use the `al`, `ax` or `eax` register per the instruction set reference.

If I accidentally wrote `reg` instead of `"eax"` in the example, would the error message be of the same quality as normal rust error messages (or at least using the same infrastructure wrt highlighting the span in question etc)?

Another concern is that it would potentially compile some of the time, when LLVM happens to allocate `eax` as its register of choice. Would that be possible?

Author: > If I accidentally wrote `reg` instead of `"eax"` in the example, would the error message be of the same quality as normal rust error messages (or at least using the same infrastructure wrt highlighting the span in question etc)?

LLVM has support for sending error messages from inline assembly back to rustc so that they can be displayed through rustc's normal error message functionality. This already works with the current `asm!` macro: https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e5d9c3c74edc6e02858ec965abfd4d98

> Another concern is that it would potentially compile some of the time, when LLVM happens to allocate `eax` as its register of choice. Would that be possible?

Yes it would sometimes compile fine and sometimes not, depending on the register selected by LLVM. There isn't much we can do about that.

Reviewer 14: More than cryptic error messages I fear inefficiency and silent errors without any error messages.

First issue...

Here's one example: https://godbolt.org/z/sT_8Qv

short foo(short input) {
    short result;
    asm("xchg %h0, %b0" : "=Q"(result) : "0"(input));
    return result;
}

How this simple function could be expressed in new inline assembler form? I couldn't see a way to do that.

Second issue is more subtle: ttps://godbolt.org/z/5d0MxL

As you can see if you would pass one-byte variable and request an "r" constraint (instead of "q" or "Q" constraint) LLVM is all to happy to make a mess out of your assembler (note how LLVM tries to return both y1 and y4 in the same eax register).

Will rustcc be able to detect and report that somehow?

Author: That seems to be a bug in LLVM. It should not allocate the *H registers for inline assembly.

Reviewer 15: The emitted code is correct, though: after the assembly block, it first stores `al` somewhere else, only then `ah` in the entire `eax` register. Technically, since you only asked for the short registers, you shouldn't have a way to e.g. zero-extend one of them and clobber the other. Should this be a strict guarantee, that you don't get the `*h` registers?

Author: Yes, the way this RFC is currently written the `*h` registers are never chosen for operands (and I don't intend to change it).

The reason why I argue that LLVM's behavior is buggy is that Clang's inline asm is designed to emulate GCC's inline asm, and GCC never allocates `*h` for register operands.

I will submit a fix to LLVM as part of the `asm!` work.

It is somewhat common that instructions have operands that are not explicitly listed in the

> assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:

Author: Implicit arguments are heavily used for syscalls, e.g.:
```rust
    asm!(
        "svc #0"
        : "={x0}" (result)
        : "{x8}" (nr as u64)
          "{x0}" (arg0)
          "{x1}" (arg1)
          "{x2}" (arg2)
          "{x3}" (arg3)
          "{x4}" (arg4)
          "{x5}" (arg5)
        : "x8", "memory"
        : "volatile"
    );
```

Something that was suggested during discussions was to lint against implicit arguments only if they don't use an explicit register.

> assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:

Author: Actually I just realized that this is already stated in the reference section:

> The compiler will lint against any operands that are not used in the template string, except for operands that specify an explicit register.

> assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:

Reviewer 10: It sounds like the request is to also lint against using positional operands like `{}` if you're also using excess operands (like those with explicit registers). That doesn't seem like an unreasonable request, as a style lint that could be disabled (or perhaps enabled).

> assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:

Reviewer 16: What about requiring such operands to have a “name” of `_`?  e.g. `asm!(“svc #0”, _ = in(“x0”) arg0, /* etc. */)`

> assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:

Reviewer 10: @comex That syntax seems confusing to me; usually `_` means "ignored", not "unnamed".

How about the following, to replace the lint text @Amanieu referenced above:

"The compiler will lint against any operands that are not substituted in the template string, except for operands that specify an explicit register. If the format string contains explicit register operands not substituted in the template string, the compiler will lint against any operands substituted using positional arguments rather than named arguments."

```rust
fn mul(a: u32, b: u32) -> u64 {
    let lo: u32;
    let hi: u32;

    unsafe {
        asm!(
            // The x86 mul instruction takes eax as an implicit input and writes
            // the 64-bit result of the multiplication to eax:edx.
            "mul {}",
            in(reg) a, in("eax") b,
            lateout("eax") lo, lateout("edx") hi
        );
    }

    hi as u64 << 32 + lo as u64
}
```

This uses the `mul` instruction to multiply two 32-bit inputs with a 64-bit result.
The only explicit operand is a register, that we fill from the variable `a`.

> The second implicit operand is the `eax` register which we fill from the variable `b`.

Reviewer 17: It took me two reads to realise that there's no previous implicit operand.
```suggestion
The second operand is implicit, and must be the `eax` register, which we fill from the variable `b`.
```

Usually this is either because we have to use a scratch register in the assembly,
or instructions modify state that we don't need to further examine.
This state is generally referred to as being "clobbered".
We need to tell the compiler about this since it may need to save and restore this state
around the inline assembly block.

```rust
let ebx: u32;
let ecx: u32;

unsafe {
    asm!(
        "cpuid",
        in("eax") 4, in("ecx") 0,
        lateout("ebx") ebx, lateout("ecx") ecx,
        lateout("eax") _, lateout("edx") _
    );
}

println!(
    "L1 Cache: {}",
    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1) * ((ebx & 0xfff) + 1) * (ecx + 1)
);
```

In the example above we use the `cpuid` instruction to get the L1 cache size.
This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.

However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.

> This can also be used with a general register class (e.g. `reg`) to obtain a scratch register for use inside the asm code.

Reviewer 17: An example of doing this would be nice, just to show how it all fits together.

Author: I added an example.

## Register template modifiers

In some cases, fine control is needed over the way a register name is formatted when inserted into the template string. This is needed when an architecture's assembly language has several names for the same register, each typically being a "view" over a subset of the register (e.g. the low 32 bits of a 64-bit register).

```rust
let mut x: u16 = 0xab;

unsafe {

>     asm!("mov {0:h} {0:b}", inout(reg_abcd) x);

Reviewer 18: I think b stands for byte and h stands for halfword, considering there's also w for word and co.

>     asm!("mov {0:h} {0:b}", inout(reg_abcd) x);

Reviewer 7: ARM uses `b` and `h`, though that looks like Intel assembly to me.

>     asm!("mov {0:h} {0:b}", inout(reg_abcd) x);

Author: We are using the same modifier flags as GCC/LLVM, defined [here](https://llvm.org/docs/LangRef.html#asm-template-argument-modifiers).

}

assert_eq!(x, 0xabab);
```

In this example, we use the `reg_abcd` register class to restrict the register allocator to the 4 legacy x86 register (`ax`, `bx`, `cx`, `dx`) of which the first two bytes can be addressed independently.

Let us assume that the register allocator has chosen to allocate `x` in the `ax` register.

> The `h` modifier will emit the register name for the high byte of that register and the `b` modifier will emit the register name for the low byte. The asm code will therefore be expanded as `mov ah, al` which copies the low byte of the value into the high byte.

Reviewer 19: Shouldn't this be a `l` instead of `b` modifier?

Author: If you look at the [LLVM reference](https://llvm.org/docs/LangRef.html#asm-template-argument-modifiers), it says that the modifier is `b` for the low byte.

- `nostack` means that the asm code does not push any data onto the stack. This allows the compiler to use optimizations such as the stack red zone on x86_64 to avoid stack pointer adjustments.

These allow the compiler to better optimize code using `asm!`, for example by eliminating pure `asm!` blocks whose outputs are not needed.

See the reference for the full list of available flags and their effects.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Inline assembler is implemented as an unsafe macro `asm!()`.
The first argument to this macro is a template string literal used to build the final assembly.
The following arguments specify input and output operands.
When required, flags are specified as the final argument.

The following ABNF specifies the general syntax:

```
dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
reg_spec := <arch specific register class> / "<arch specific register name>"

> operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"

Reviewer 20: > `operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"`

Should the `expr`s in this line be `ident`s, or do we really support arbitrary expressions in all these locations?

Author: We do want these to be full expressions. This allows for example using `my_struct.field` as an asm operand.

reg_operand := dir_spec "(" reg_spec ")" operand_expr
operand := reg_operand / "imm" const_expr / "sym" path
flag := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
flags := "flags(" flag *["," flag] ")"
asm := "asm!(" format_string *("," [ident "="] operand) ["," flags] ")"
```

[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax

## Template string

The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.

> The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.

Reviewer 20: > The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.

Can we see an example of Intel syntax being used with this macro?

Author: All examples in this RFC use Intel syntax.

Reviewer 10: In order to aid in mechanical translation of code, could we expose the appropriate syntax flags here, so that people *can* ask for AT&T syntax on x86, for instance? (That would make it easy for others to implement an `asm_att!` macro, for instance.)

And, of course, the alternatives section should mention the possibility of using AT&T syntax on all platforms and providing a flag for Intel syntax, in which case people could easily implement an `asm_intel!` macro for Intel syntax.

Author: This is tricky to support reliably. In particular, GCC requires that all asm code in a compilation unit use the same syntax, so this would at least exclude the possibility of inline asm support with a GCC backend.

I would prefer if we simply chose a single asm syntax and stuck with it. Also note that this only affects x86 which has 2 asm syntaxes. Every other architecture only has a single standardized asm syntax.

Reviewer 7: Could it be a crate level setting, since each crate is a single translation unit?

Author: I don't think that a crate-level setting will be enough if you take LTO into account. And even then, a crate-level setting won't work with inline functions.

Reviewer 16: No need for that.  You can switch syntaxes on-the-fly using assembler directives, `.att_syntax` and `.intel_syntax noprefix`: [example](https://gcc.godbolt.org/z/d28KDo).

Author: The issue is more with the register placeholders, since GCC needs to know whether to emit `eax` or `%eax`.

Reviewer 16: Oh... good point.  For the record, this only affects a hypothetical GCC backend, not LLVM, which supports `inteldialect` per asm block.

I suppose such a backend could always tell GCC to compile in Intel mode, and then just add the `%`s and `$`s itself.

Reviewer 10: @Amanieu GCC supports prefixed registers even when in Intel syntax mode, and even with `noprefix`; `noprefix` just makes the prefixes optional. So you can (and should) always emit prefixes like `%` and `$`, always leave GCC in AT&T mode, and just wrap assembly blocks that use Intel syntax in `.intel_syntax noprefix` and `.att_syntax`.

  - An underscore (`_`) may be specified instead of an expression for `<out expr>`, which will cause the contents of the register to be discarded at the end of the asm code (effectively acting as a clobber).
  - `<in expr>` and `<out expr>` may have different types.
* `inlateout(<reg>) <expr>` / `inlateout(<reg>) <in expr> => <out expr>`
  - Identical to `inout` except that the register allocator can reuse a register allocated to an `in` (this can happen if the compiler knows the `in` has the same initial value as the `inlateout`).
  - You should only write to the register after all inputs are read, otherwise you may clobber an input.
* `imm <expr>`
  - `<expr>` must be an integer or floating-point constant expression.
  - The value of the expression is formatted as a string and substituted directly into the asm template string.
* `sym <path>`
  - `<path>` must refer to a `fn` or `static` defined in the current crate.
  - A mangled symbol name referring to the item is substituted into the asm template string.
  - The substituted string does not include any modifiers (e.g. GOT, PLT, relocations, etc).

## Register operands

> Input and output operands can be specified either as an explicit register or as a register class from which the register allocator can select a register. Explicit registers are specified as string literals (e.g. `"eax"`) while register classes are specified as raw identifiers (e.g. `reg`).

Reviewer 3: Nit: I think "raw identifier" specifically means something like [`r#match`](https://doc.rust-lang.org/edition-guide/rust-2018/module-system/raw-identifiers.html). So I would just drop "raw" here.

Different registers classes have different constraints on which Rust types they allow. For example, `reg` generally only allows integers and pointers, but not floats or SIMD vectors.

If a value is of a smaller size than the register it is allocated in then the upper bits of that register will have an undefined value for inputs and will be ignored for outputs. It is a compile-time error for a value to be of a larger size than the register it is allocated in.

Here is the list of currently supported register classes:

| Architecture | Register class | Registers | LLVM constraint code | Allowed types |
| ------------ | -------------- | --------- | ----- | ------------- |
| x86 | `reg` | `ax`, `bx`, `cx`, `dx`, `si`, `di`, `r[8-15]` (x86-64 only) | `r` | `i8`, `i16`, `i32`, `i64` (x86-64 only) |
| x86 | `reg_abcd` | `ax`, `bx`, `cx`, `dx` | `Q` | `i8`, `i16`, `i32`, `i64` (x86-64 only) |

> | x86 | `vreg` | `xmm[0-7]` (x86) `xmm[0-15]` (x86-64) | `x` | `i32`, `i64`, `f32`, `f64`, `v128`, `v256`, `v512` |

Reviewer 20: Does `v128` etc refer to some proposal for standardized SIMD types?

Author: It's just a notation used in this RFC, see the notes below the table. Regarding SIMD types, this basically means that you can use `__m128` as an asm operand.

Reviewer 10: We either need to support `i128` here, or we need to document why LLVM makes that difficult (and report it as an issue on LLVM).

Author: `i128` doesn't fit in one register though. If you want to pass it as a register pair then you should split it yourself.

Reviewer 21: @Amanieu isn't that target-dependent? e.g. on RV128 you should be able to use `i128` as a register just fine. OTOH, `x86_64` does not have arithmetic registers for 128-bit integers and stores `i128` in two 64-bit registers instead, so there you need to split it when passing it to/from an assembly block (e.g. as the result of `mulxq`, which "returns" two 64-bit integers).

| x86 (AVX-512) | `kreg` | `k[1-7]` | `Yk` | `i16`, `i32`, `i64` |
| AArch64 | `reg` | `x[0-28]`, `x30` | `r` | `i8`, `i16`, `i32`, `i64` |
| AArch64 | `vreg` | `v[0-31]` | `w` | `i8`, `i16`, `i32`, `i64`, `f32`, `f64`, `v64`, `v128` |
| AArch64 | `vreg_low` | `v[0-15]` | `x` | `i8`, `i16`, `i32`, `i64`, `f32`, `f64`, `v64`, `v128` |
| AArch64 | `vreg_low8` | `v[0-7]` | `y` | `i8`, `i16`, `i32`, `i64`, `f32`, `f64`, `v64`, `v128` |
| ARM | `reg` | `r[0-r10]`, `r12`, `r14` | `r` | `i8`, `i16`, `i32` |
| ARM | `vreg` | `s[0-31]`, `d[0-31]`, `q[0-15]` | `w` | `f32`, `f64`, `v64`, `v128` |
| ARM | `vreg_low` | `s[0-31]`, `d[0-15]`, `q[0-7]` | `t` | `f32`, `f64`, `v64`, `v128` |
| ARM | `vreg_low8` | `s[0-15]`, `d[0-d]`, `q[0-3]` | `x` | `f32`, `f64`, `v64`, `v128` |
| RISC-V | `reg` | `x1`, `x[5-7]`, `x[9-31]` | `r` | `i8`, `i16`, `i32`, `i64` (RV64 only) |
| RISC-V | `vreg` | `f[0-31]` | `f` | `f32`, `f64` |

> Notes on allowed types:
> - Pointers and references are allowed where the equivalent integer type is allowed.
> - `iLEN` refers to both sized and unsized integer types. It also implicitly includes `isize` and `usize` where the length matches.
> - Fat pointers are not allowed.

> > - `vLEN` refers to a SIMD vector that is `LEN` bits wide.

Reviewer 12: What is `iLEN` and `vLEN` here? I do not see it referred to anywhere else.

Author: `i8` `i16` `i32` `i64` in the table just above.

| ARM | `r12` | `ip` |
| ARM | `r13` | `sp` |
| ARM | `r14` | `lr` |
| ARM | `r15` | `pc` |
| RISC-V | `x0` | `zero` |
| RISC-V | `x1` | `ra` |
| RISC-V | `x2` | `sp` |
| RISC-V | `x3` | `gp` |
| RISC-V | `x4` | `tp` |
| RISC-V | `x[5-7]` | `t[0-2]` |
| RISC-V | `x8` | `fp`, `s0` |
| RISC-V | `x9` | `s1` |
| RISC-V | `x[10-17]` | `a[0-7]` |
| RISC-V | `x[18-27]` | `s[2-11]` |
| RISC-V | `x[28-31]` | `t[3-6]` |
| RISC-V | `f[0-7]` | `ft[0-7]` |
| RISC-V | `f[8-9]` | `fs[0-1]` |
| RISC-V | `f[10-17]` | `fa[0-7]` |
| RISC-V | `f[18-27]` | `fs[2-11]` |
| RISC-V | `f[28-31]` | `ft[8-11]` |

Some registers cannot be used for input or output operands:

| Architecture | Unsupported register | Reason |
| ------------ | -------------------- | ------ |
| All | `sp` | The stack pointer must be restored to its original value at the end of an asm code block. |
| All | `bp` (x86), `r11` (ARM), `x29` (AArch64), `x8` (RISC-V) | The frame pointer cannot be used as an input or output. |
| x86 | `ah`, `bh`, `ch`, `dh` | These are poorly supported by compiler backends. Use 16-bit register views (e.g. `ax`) instead. |
| x86 | `k0` | This is a constant zero register which can't be modified. |

> | x86 | `ip` | This is the program counter, not a real register. |

Reviewer 12: Why does `ip` have aliases if it is unusable?

Author: So that the compiler can provide better error messages ("unknown register" vs "disallowed register").

| AArch64 | `xzr` | This is a constant zero register which can't be modified. |
| ARM | `pc` | This is the program counter, not a real register. |
| RISC-V | `x0` | This is a constant zero register which can't be modified. |

> | RISC-V | `gp`, `tp` | These registers are reserved and cannot be used as inputs or outputs. |

Reviewer 12: I am confused about this list. It is written as though it is an exhaustive list, and even goes so far as to mention a register only used in a fairly recent instruction set extension (`k0` for x86).

Wouldn't it make more sense to just list a few examples here and keep an up-to-date list in the documentation?

Author: The documentation for inline assembly will most likely be a verbatim copy of the contents of this RFC.

## Flags

Flags are used to further influence the behavior of the inline assembly block.
Currently the following flags are defined:
- `pure`: The `asm` block has no side effects, and its outputs depend only on its direct inputs (i.e. the values themselves, not what they point to). This allows the compiler to execute the `asm` block fewer times than specified in the program (e.g. by hoisting it out of a loop) or even eliminate it entirely if the outputs are not used. A warning is emitted if this flag is used on an `asm` with no outputs.
- `nomem`: The `asm` blocks does not read or write to any memory. This allows the compiler to cache the values of modified global variables in registers across the `asm` block since it knows that they are not read or written to by the `asm`.
- `readonly`: The `asm` block does not write to any memory. This allows the compiler to cache the values of unmodified global variables in registers across the `asm` block since it knows that they are not written to by the `asm`.
- `preserves_flags`: The `asm` block does not modify the flags register (defined below). This allows the compiler to avoid recomputing the condition flags after the `asm` block.

> - `noreturn`: The `asm` block never returns, and its return type is defined as `!` (never). Behavior is undefined if execution falls through past the end of the asm code.

Reviewer 20: > its return type is defined as `!` (never)

Ordinarily does `asm!` have a return type of `()`?

Reviewer 5: Also, is there any difference between using `noreturn` and putting a call to [`unreachable_unchecked`](https://doc.rust-lang.org/std/hint/fn.unreachable_unchecked.html) after the `asm!` block?

Author: Yes, `asm!` has a return type of `()`.

> Also, is there any difference between using noreturn and putting a call to unreachable_unchecked after the asm! block?

If you look at the "Mapping to LLVM IR" section, you'll see that's exactly what it does.

Reviewer 5: Yes, I saw this:

> If the `noreturn` flag is set then an `unreachable` LLVM instruction is inserted after the asm invocation.

But I think `unreachable_unchecked` will result in the same code (modulo inlining, etc.). Do you think this has a performance impact, or is there another reason other than brevity to prefer the `noreturn` modifier?

Author: I added the `noreturn` flag because I feel that it makes it more explicit that the asm never returns.

Reviewer 13: There’s another reason to want this here: naked functions. Those are only allowed to have a single asm statement - anything else is UB. This means that currently, you can’t have a naked function that returns !, which is pretty inconvenient.

With this noreturn flag, the problem is neatly solved.

Reviewer 22: Umm... can't you just stick a panic at the end of the function to make it return never?

Reviewer 13: Naked functions can't have a panic at the end of the function. Again, naked functions are **only** allowed to have a single asm statement - nothing else.

Reviewer 19: What is the advantage of a naked function over a `global_asm!` block?

Author: The main advantage is that the compiler handles name mangling for you.

Reviewer 22: IMO, the main advantage is that it is inline, which is much easier to read than jumping to another file and makes it easier to deal with things like accessing globals or modifying rust variables.

@roblabla Is this documented somewhere? I've used a lot of naked functions with function calls etc, and it seems to work fine as long as your asm sets up the stack properly.

Reviewer 22: Btw, I'm glad to discuss naked asm further, but perhaps we should do it in the tracking issue instead?

Reviewer 19: > IMO, the main advantage is that it is inline, which is much easier to read than jumping to another file

You don't have to place `global_asm!();` in a different file.

>  and makes it easier to deal with things like accessing globals or modifying rust variables.

:+1:

Reviewer 13: @mark-i-m 

>it is generally unsafe to write anything but inline assembly inside a naked function.  The [LLVM language reference](http://llvm.org/docs/LangRef.html#function-attributes) describes this feature as having "very system-specific consequences", which the programmer must
be aware of.

https://github.com/rust-lang/rfcs/pull/1201/files#diff-e089b5aa9b57392ee8a5efa292856d1eR74

There is [an open RFC](https://github.com/rust-lang/rfcs/pull/2774/files) to amend 1201 to explicitly deny various common misuses of it. Let's move naked function discussion there.

> - `noreturn`: The `asm` block never returns, and its return type is defined as `!` (never). Behavior is undefined if execution falls through past the end of the asm code.

Reviewer 3: I was actually confused by the name `noreturn`. I initially assumed this meant that there is no `ret` instruction (or similar) in the assembly. How about using `diverging` or `diverge` as the name instead?

Author: I feel that `noreturn` expresses this better: execution never returns from the asm block.

- `nostack`: The `asm` block does not push data to the stack, or write to the stack red-zone (if supported by the target). If this flag is *not* used then the stack pointer is guaranteed to be suitably aligned (according to the target ABI) for a function call.

> The `nomem` and `readonly` flags are mutually exclusive: it is an error to specify both. Specifying `pure` on an asm block with no outputs is linted against since such a block will be optimized away to nothing.

Reviewer 20: > it is an error to specify both

A compiler error?

Author: Yes.

the variable name, which makes this syntax overall harder to read.

```rust
fn mul(a: u32, b: u32) -> u64 {
    let (lo, hi) = unsafe {
        asm!("mul {}", in(reg) a, in("eax") b, lateout("eax"), lateout("edx"))
    };

    hi as u64 << 32 + lo as u64
}
```

# Prior art
[prior-art]: #prior-art

## GCC inline assembly

The proposed syntax is very similar to GCC's inline assembly in that it is based on string substitution while leaving actual interpretation of the final string to the assembler. However GCC uses poorly documented single-letter constraint codes and template modifiers. Clang tries to emulate GCC's behavior, but there are still several cases where its behavior differs from GCC's.

The main reason why this is so complicated is that GCC's inline assembly basically exports the raw internals of GCC's register allocator. This has resulted in many internal constraint codes and modifiers being widely used, despite them being completely undocumented.

## D & MSVC inline assembly

See the section [above][dsl].

# Unresolved questions
[unresolved-questions]: #unresolved-questions

None

# Future possibilities
[future-possibilities]: #future-possibilities

## Flag outputs

> GCC supports a special type of output which allows an asm block to return a `bool` encoded in the condition flags register. This allows the compiler to branch directly on the condition flag instead of materializing the condition as a `bool`.

Reviewer 23: This would be super useful for Intel ADCX/ADOX instructions for multi-precision integer math. I have an example of a simple assembly multiplication routine [here](https://github.com/0xProject/OpenZKP/blob/6e59a51e39eaa33331d0860b6b60d7a70a19f709/algebra/u256/src/algorithms/assembly.rs#L86) It would be useful for me if I could specify the carry and overflow flags as inputs/outputs.

This is in line with what the LLVM instrinsics for these instructions do. See [`core::arch::x86_64::_addcarryx_u64`](https://doc.rust-lang.org/core/arch/x86_64/fn._addcarryx_u64.html).

Unfortunately, again it seems like LLVM just ignores it. As mentioned in https://github.com/rust-lang/stdarch/issues/666#issuecomment-485339414, there is an open LLVM issue and it does not seem easy:

> The X86 backend isn't currently set up to model the C flag and O flag separately. We model all of the flags as one register. Because of this we can't interleave the flag dependencies. We would need to do something about that before it makes sense to implement _addcarryx_u64 as anything other than plain adc.
> -- https://bugs.llvm.org/show_bug.cgi?id=41546#c1

We can support this in the future with a special output operand type.

## `asm goto`

GCC supports passing C labels (the ones used with `goto`) to an inline asm block, with an indication that the asm code may jump directly to one of these labels instead of leaving the asm block normally.

This could be supported by allowing code blocks to be specified as operand types. The following code will print `a` if the input value is `42`, or print `b` otherwise.

```rust
asm!("cmp {}, 42; jeq {}",
    in(reg) val,
    label { println!("a"); },
    fallthrough { println!("b"); }
);
```

## Unique ID per `asm`

GCC supports `%=` which generates a unique identifier per instance of an asm block. This is guaranteed to be unique even if the asm block is duplicated (e.g. because of inlining).

We can support this in the future with a special operand type.

## `imm` and `sym` for `global_asm!`

The `global_asm!` macro could be extended to support `imm` and `sym` operands since those can be resolved by simple string substitution. Symbols used in `global_asm!` will be marked as `#[used]` to ensure that they are not optimized away by the compiler.

## Memory operands

> We could support `mem` as an alternative to specifying a register class which would leave the operand in memory and instead produce a memory address when inserted into the asm string. This would allow generating more efficient code by taking advantage of addressing modes instead of using an intermediate register to hold the computed address.

Reviewer 23: It's very common to have instructions (like `add` or `mov`) that accept inputs as both registers or memory. Specifying this as a `rm` constraint allows the compiler to do better register allocation/spilling.

That being said, LLVM currently ignores any freedom in the constraints and always picks memory:

> Thus, it simply tries to make a choice that’s most likely to compile, not one that will be optimal performance. (e.g., given “rm”, it’ll always choose to use memory, not registers).
> - [Section 'Constraint codes'](https://releases.llvm.org/9.0.0/docs/LangRef.html#module-level-inline-assembly)

Author: Also LLVM only really supports memory addressing modes for inline asm on x86. On ARM/AArch64 it will just perform the address calculation and put the result in a register.

    );
}

println!(
    "L1 Cache: {}",
    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1) * ((ebx & 0xfff) + 1) * (ecx + 1)
);
```

In the example above we use the `cpuid` instruction to get the L1 cache size.
This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.

However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.

This can also be used with a general register class (e.g. `reg`) to obtain a scratch register for use inside the asm code.

## Register template modifiers

In some cases, fine control is needed over the way a register name is formatted when inserted into the template string. This is needed when an architecture's assembly language has several names for the same register, each typically being a "view" over a subset of the register (e.g. the low 32 bits of a 64-bit register).

```rust
let mut x: u16 = 0xab;

unsafe {

>     asm!("mov {0:h} {0:b}", inout(reg_abcd) x);

Reviewer 24: If I'm reading this correctly. This code transforms 0xab to 0xabab. But the description states 0x0a0b to 0x0b0b.

}

assert_eq!(x, 0xabab);
```

In this example, we use the `reg_abcd` register class to restrict the register allocator to the 4 legacy x86 register (`ax`, `bx`, `cx`, `dx`) of which the first two bytes can be addressed independently.

Let us assume that the register allocator has chosen to allocate `x` in the `ax` register.
The `h` modifier will emit the register name for the high byte of that register and the `b` modifier will emit the register name for the low byte. The asm code will therefore be expanded as `mov ah, al` which copies the low byte of the value into the high byte.

> ## Flags

Reviewer 17: English here is challenging, because there are two meanings of "flags" to juggle:

1. The CPU flags, as contained in (e.g.) the RFLAGS register on x86_64 or the PSR on ARM.
2. Flags that change the behaviour of the `asm!` block.

Is there a better name we could use for meaning 2 here, instead of flags, to avoid confusion?

Reviewer 17: Perhaps `modifiers` or `options` here, instead of `flags`, to avoid confusion between the two meanings.

Author: This is a good point which I kinda agree with. However neither `modifiers` nor `options` feel like very good fits. One alternative would be to eliminate the `flags()` keyword entirely and just have flags appended to the operand list.

Reviewer 12: Attributes perhaps?

Reviewer 10: Excellent suggestion! Changing this to "options" sounds good. ("modifiers" can have similar ambiguity.)

Reviewer 10: @Amanieu Appending the individual options to the operand list seems reasonable; that would make the use of such options less verbose. That would mean options would have to appear distinct from other things that might appear in the operand list, but that seems like a perfectly reasonable requirement.

Reviewer 21: Currently the `flags`/`options` control the "effects" that the assembly block has, so we can also call it `effects(nomem, nostack, noflags, noregs)` to state that the assembly block does not have effects that modify "x" - where `pure` and `readonly` are shorthands for some of the effects.

Inline assembler is implemented as an unsafe macro `asm!()`.
The first argument to this macro is a template string literal used to build the final assembly.
The following arguments specify input and output operands.
When required, flags are specified as the final argument.

The following ABNF specifies the general syntax:

```
dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
reg_spec := <arch specific register class> / "<arch specific register name>"
operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
reg_operand := dir_spec "(" reg_spec ")" operand_expr
operand := reg_operand / "imm" const_expr / "sym" path
flag := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
flags := "flags(" flag *["," flag] ")"
asm := "asm!(" format_string *("," [ident "="] operand) ["," flags] ")"
```

[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax

## Template string

The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.

> The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.

Reviewer 25: Does this imply that `x86` will use the format of `mnemonic	destination, source` (like Intel Syntax) while every other platform will use `mnemonic	source, destination` (like AT&T/GAS)? Because that'll extremely hurt portability/readability.

Author: Every other architecture already uses `mnemonic destination, source`. x86 is the exception with AT&T syntax that reverses this order.

Reviewer 25: Okay, so every platform is standardized to `mnemonic dst, src`?

This is a bit confusing, as I'm not sure if the Rust-Source is being specified, or the IR which is being transferred to the compiler. 

So does this imply that on _non-x86_ platforms `rustc` will be responsible for re-ordering `mnemonic` arguments to the backend?

Author: There is no reordering needed, all non-x86 platforms already use the correct ordering. The Intel/ATT syntax split is only a thing on x86.

| x86 | `vreg_evex` | `xmm[0-31]` (AVX-512, otherwise same as `vreg`) | `v` | `i32`, `i64`, `f32`, `f64`, `v128`, `v256`, `v512` |
| x86 (AVX-512) | `kreg` | `k[1-7]` | `Yk` | `i16`, `i32`, `i64` |
| AArch64 | `reg` | `x[0-28]`, `x30` | `r` | `i8`, `i16`, `i32`, `i64` |
| AArch64 | `vreg` | `v[0-31]` | `w` | `i8`, `i16`, `i32`, `i64`, `f32`, `f64`, `v64`, `v128` |
| AArch64 | `vreg_low` | `v[0-15]` | `x` | `i8`, `i16`, `i32`, `i64`, `f32`, `f64`, `v64`, `v128` |
| AArch64 | `vreg_low8` | `v[0-7]` | `y` | `i8`, `i16`, `i32`, `i64`, `f32`, `f64`, `v64`, `v128` |
| ARM | `reg` | `r[0-r10]`, `r12`, `r14` | `r` | `i8`, `i16`, `i32` |
| ARM | `vreg` | `s[0-31]`, `d[0-31]`, `q[0-15]` | `w` | `f32`, `f64`, `v64`, `v128` |
| ARM | `vreg_low` | `s[0-31]`, `d[0-15]`, `q[0-7]` | `t` | `f32`, `f64`, `v64`, `v128` |
| ARM | `vreg_low8` | `s[0-15]`, `d[0-d]`, `q[0-3]` | `x` | `f32`, `f64`, `v64`, `v128` |
| RISC-V | `reg` | `x1`, `x[5-7]`, `x[9-31]` | `r` | `i8`, `i16`, `i32`, `i64` (RV64 only) |
| RISC-V | `vreg` | `f[0-31]` | `f` | `f32`, `f64` |

> Notes on allowed types:
> - Pointers and references are allowed where the equivalent integer type is allowed.

> > - `iLEN` refers to both sized and unsized integer types. It also implicitly includes `isize` and `usize` where the length matches.

Reviewer 26: Do you mean "signed and unsigned" instead of "sized and unsized?"

Reviewer 12: You fixed it to "sized and unsigned" instead of "signed and unsigned".

The higher 32 bits are stored in `edx` from which we fill the variable `hi`.

Note that `lateout` must be used for `eax` here since we are specifying the same register as both an input and an output.

## Clobbered registers

In many cases inline assembly will modify state that is not needed as an output.
Usually this is either because we have to use a scratch register in the assembly,
or instructions modify state that we don't need to further examine.
This state is generally referred to as being "clobbered".
We need to tell the compiler about this since it may need to save and restore this state
around the inline assembly block.

```rust
let ebx: u32;
let ecx: u32;

unsafe {
    asm!(
        "cpuid",
        in("eax") 4, in("ecx") 0,
        lateout("ebx") ebx, lateout("ecx") ecx,
        lateout("eax") _, lateout("edx") _
    );
}

println!(
    "L1 Cache: {}",
    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1) * ((ebx & 0xfff) + 1) * (ecx + 1)
);
```

In the example above we use the `cpuid` instruction to get the L1 cache size.
This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.

> However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.

Reviewer 27: I'd prefer if things like clobbering could be motivated and specified in terms of the constraints that are put *on the assembly*, not on the compiler. I gather people writing inline assembly are used to thinking very much in terms of compiler details such as register allocation, but I'd prefer if any such remarks would be non-normative notes.

For example, I'd say that the motivation for explicit declaration of clobbered registers is that assembly may not access (read nor write) any register not declared in its inputs and outputs. So, the `cpuid` example without the `lateout("eax") _, lateout("edx") _` is UB due to violating those constraints. That motivates clobbering without ever mentioning what the compiler ends up doing with this information. The fact that the compiler has to save and restore registers is a compiler implementation detail that follows from this description, but it should not be normative.

Phrasing things in terms of constraints on the compiler is akin to specifying a language feature by the set of allowed optimizations: it leaves us with no way of understanding the feature as a language construct on its own right, independent of compilation or any such thing. This is the same concern as trying to specify `black_box` as "inhibiting optimizations" -- neither optimizations nor things like register save/restore should occur in a language spec, IMO.

From what I can tell, basically everything in here *can* be described in the abstract way I propose, which makes me very happy! As a formal methods person, inline assembly is something like my Kryptonite, but this RFC is very well-written and I felt I could mostly follow it. :D   It's just a matter of wording, I think.  The one point I am less sure about is the `lateout` thing... is it enough to say "if any input register gets read after this output was written to, that's UB"?

Author: This part is in the guide-level explanation, which tries to be a gentle (ish) introduction to inline assembly rather than a proper specification. In this particular case, it helps to explain to users why indicating proper clobbers is important. The reference-level explanation is more rigorous in specifying exactly what is/isn't UB.

Now regarding the `lateout` thing... I wouldn't say that behavior is undefined if you misuse it, just that you may end up (depending on the register allocator's choices) accidentally overwriting one of your inputs. The problem is entirely confined to the asm code and does not cause any UB and the Rust abstract machine or LLVM level. Hence the looser wording ("should" instead of "must").

Reviewer 27: > This part is in the guide-level explanation, which tries to be a gentle (ish) introduction to inline assembly rather than a proper specification. In this particular case, it helps to explain to users why indicating proper clobbers is important.

I agree that it helps to motivate why something is UB. But even for the gentle introduction, I think it would help to frame this as "that code is UB (and here's why), and here's how we add these clobber annotations to fix that". It puts the reader into the right frame of mind. Otherwise, we risk educating people to primarily think about the compiler's choices when writing such code.

The way things stand right now, the guide does not ever say anything about UB. That is quite unusual for Rust where we are usually very explicit, even in guide-level material, about what the rules are to avoid UB.

> The reference-level explanation is more rigorous in specifying exactly what is/isn't UB.

The term "UB"/"Undefined Behavior" does not even occur in the RFC, though. So, what you are saying is not entirely accurate.
I also looked for the clause that says that reading or writing any register not explicitly specified is UB, and could not find that either.

> Now regarding the lateout thing... I wouldn't say that behavior is undefined if you misuse it, just that you may end up (depending on the register allocator's choices) accidentally overwriting one of your inputs.

If misusing lateout is not UB, we have to specify what happens in that case. So, what does that spec look like? "Non-deterministically, writing to a lateout register might overwrite an input register"?

> Hence the looser wording ("should" instead of "must").

It is far from clear that this is a deliberate choice. I am not sure what "should" means in a spec. Either behavior is UB or it is specified. The spec isn't really the right place for recommendations or advice.

Author: For lateout, the specification is:

> the register allocator can reuse a register allocated to an `in`.

Everything else follows as a consequence of that: if you write to the output, you may overwrite one of your inputs.

Reviewer 28: > Now regarding the lateout thing... I wouldn't say that behavior is undefined if you misuse it, just that you may end up (depending on the register allocator's choices) accidentally overwriting one of your inputs. The problem is entirely confined to the asm code and does not cause any UB and the Rust abstract machine or LLVM level. Hence the looser wording ("should" instead of "must").

On some level it depends on whether we define this as undefined behavior. I'd rather do so, because breaking this invariant can trivially cause undefined behavior in otherwise well-defined (assembly) code.
Consider for example:
```rust
let xs = [1u32, 2];
let mut sum;
asm!("
    mov {sum41}, 41
    sum_loop:
        add {sum41}, [{ptr} + ecx*4 - 1]
    loop sum_loop
    ",
    sum41 = lateout(reg) sum,
    ptr = in(reg) xs.as_ptr(),
    in("ecx") xs.len(),
);
assert_eq!(sum, 44);
```
Whenever the `lateout` shadows either `in` this causes out-of-bounds accesses.

Reviewer 29: Is out-of-bounds access within an asm block even considered undefined behavior in "the compiler may optimize assuming this code is never executed at all" meaning? I would expect a much weaker assumption that the compiler can just assume that the out-of-bounds accesses don't touch anything rustc knows/cares about.

> For example, I'd say that the motivation for explicit declaration of clobbered registers is that assembly may not access (read nor write) any register not declared in its inputs and outputs.

This is way too strong. In x86, the NOP instruction literally means "swap the contents of the eax register with the contents of eax", but it would be silly to require any assembly block with a nop in it to mark eax as a clobber. In general reading registers is always harmless and writing them only matters if their value isn't restored later, so it seems like it should be enough to say that it is undefined behavior to modify (and not subsequently restore) a register unless it is listed in an output/clobber.

>  The one point I am less sure about is the lateout thing... is it enough to say "if any input register gets read after this output was written to, that's UB"?

Why do you need to say anything? I'd expect all the register constraints to be in terms of which assignments can alias with which other assignments, and lateout would have the property of needing a register which is distinct from any other output register or clobber, but possibly aliasing an input register. Reasoning about the assembly block would then come down to considering how it behaves under each possible aliasing combination.

Reviewer 13: > For example, I'd say that the motivation for explicit declaration of clobbered registers is that assembly may not access (read nor write) any register not declared in its inputs and outputs.

This is wrong though. For instance, the following piece of inline asm does not clobber eax:

```rust
asm!("push eax; mov eax, 4; pop eax");
```

Reviewer 21: > Is out-of-bounds access within an asm block even considered undefined behavior 

No, this is not undefined behavior, e.g., 

```rust
let ptr: *mut f32 = GlobalAlloc::alloc(Layout::new(8, 16)?) as *mut _;
assert!(ptr as usize % 16 == 0); // aligned to a 16 byte boundary
let mut arr = [MaybeUninit<f32>; 4];
asm!("
        simd_16byte_load xmm0, {0}
        simd_16byte_write {1}, xmm0
    ", 
    in(ptr), out(&arr));
```

allocates 8 bytes on the heap at a 16 byte aligned boundary, and then uses a SIMD load to read 16 byte from the 8-byte wide allocation, reading the last 8-bytes out-of-bounds within the assembly block, and that's ok.

What's UB is an `asm!` block that has a side-effect but does not state it, e.g., an `asm!` block that modifies a register, but does not mark it as being an output or a clobber, since the compiler will generate code under the assumption that this register is not modified, yet that assumption would be incorrect.

Reviewer 27: > This is way too strong. In x86, the NOP instruction literally means "swap the contents of the eax register with the contents of eax", but it would be silly to require any assembly block with a nop in it to mark eax as a clobber. In general reading registers is always harmless and writing them only matters if their value isn't restored later, so it seems like it should be enough to say that it is undefined behavior to modify (and not subsequently restore) a register unless it is listed in an output/clobber.

Thanks for adding that, I was not aware of this.

@gnzlbg's definition makes sense to me.

> This subtly different definition means that an assembly block can actually read from almost any register, including those that are not specified as inputs (e.g. like flags registers), because doing so is not a side-effect. I say almost any, because reading from some registers can be a side-effect (i.e. can modify the value of that or other registers).

It probably also goes without saying that there are no guarantees whatsoever for the value that is being read from such registers.

Reviewer 15: > ("Read" is a kind of effect though, in the sense that entirely pure code ["mathematical functions"] must not even read from mutable things. It's just a harmless effect in many cases. I am not sure how much bearing this fact has on the discussion at hand, but I get a bit worried every time someone claims that reads don't have side-effects.)

I think this is a software (compiler) vs hardware thing. If you read in source code, the compiler has to make sure there's a value there for you to read, which affects the code it generates, meaning there were side-effects to that read - and if the thing you were reading is mutable, this might visibly change what's written there (conversely, if you're a pure function then it might not make sure that you have something to read and then UB). At the hardware level, there's always something to read at that register, even if you can't guarantee what it is, so no extra effort needs to be made. A pure Rust function could have inline asm with the pure marker, and reading from a random register should still be no problem (although, of course, odd). Maybe otherwise put, in the "memory-model" of the cpu-registers, read really isn't an effect (except for the special registers where it is). For actual physical memory, that's probably more delicate.

The `h` modifier will emit the register name for the high byte of that register and the `b` modifier will emit the register name for the low byte. The asm code will therefore be expanded as `mov ah, al` which copies the low byte of the value into the high byte.

## Flags

By default, an inline assembly block is treated the same way as an external FFI function call with a custom calling convention: it may read/write memory, have observable side effects, etc. However in many cases, it is desirable to give the compiler more information about what the assembly code is actually doing so that it can optimize better.

Let's take our previous example of an `add` instruction:

```rust
let mut a = 4;
let b = 4;
unsafe {
    asm!(
        "add {0}, {1}",
        inlateout(reg) a, in(reg) b,
        flags(pure, nomem, nostack)
    );
}
assert_eq!(a, 8);
```

Flags can be provided as an optional final argument to the `asm!` macro. We specified three flags here:
- `pure` means that the asm code has no observable side effects and that its output depends only on its inputs. This allows the compiler optimizer to call the inline asm fewer times or even eliminate it entirely.

> - `nomem` means that the asm code does not read or write to memory. By default the compiler will assume that inline assembly can read or write any memory address that is accessible to it (e.g. through a pointer passed as an operand, or a global).

Reviewer 27: > By default the compiler will assume that inline assembly can read or write any memory address that is accessible to it

It might be worth clarifying that "accessible" here means "accessible according to Rust's aliasing rules". I.e., this is not allowed to perform any kind of access that unsafe Rust wouldn't also be permitted to do. In particular, if we end up adopting a variant of Stacked Borrows, all memory accesses performed by assembly are subject to the usual Stacked Borrows rules for memory accesses.

Reviewer 27: Also, the wording should be careful not to accidentally make `black_box`-style guarantees. By default, the assembly *may* read or write any accessibly memory address -- but that does not *guarantee* that the compiler will consider all that memory clobbered, it just means that the compiler would have to do some smart analysis of the assembly to figure out what did and what did not get clobbered.

Reviewer 7: So, can wr also put a black_box flag then?

Reviewer 27: I have no idea what that flag would mean. But the goal is that it must not be possible to implement a `black_box` with this that is any stronger than [this one](https://doc.rust-lang.org/nightly/core/hint/fn.black_box.html):

> Note however, that black_box is only (and can only be) provided on a "best-effort" basis. The extent to which it can block optimisations may vary depending upon the platform and code-gen backend used. Programs cannot rely on black_box for correctness in any way.

Author: Basically the rule of thumb is that inline asm code is allowed to call arbitrary Rust functions. The memory access rules derive from that. Even analysis of the asm by the compiler cannot rule this out:
- The asm could contain a `call` instruction.
- The asm could contain an `int3` instruction which traps and invokes a signal handler. The signal handler may read/write arbitrary memory addresses (within the usual Rust rules).
- The asm could contain a load/store instruction which segfaults. The segfault is "fixed up" by the signal handler and execution is resumed.

Since asm also guarantees that the exact instructions specified will be the ones executed, the most a compiler could do by analyzing inline asm code is to use it as a hint for instruction scheduling around the asm code.

Reviewer 27: All this is only true in general. In specific cases, e.g. a `nop` asm block, the compiler knows exactly that no memory is read from or written to, and it can exploit that when optimizing the surrounding code.

Reviewer 16: I strongly believe that the compiler should be forbidden from analyzing assembly blocks.  So it would indeed be a stronger guarantee than `hint::black_box`.

How can this be justified?  Well, I was admittedly against limiting `black_box` to best-effort in the first place.  But a significant difference that `asm!` is architecture-specific and explicitly tied to the architectural-level machine model, whereas `black_box` is a vaguer promise to prohibit 'optimizations'.  I thought `black_box`'s semantics should be defined as, more or less, its implementation: translated to the new syntax,
```rust
pub fn black_box<T>(dummy: T) -> T {
    unsafe {
        asm!("", inout reg(dummy));
        return dummy;
    }
}
```
But admittedly, that may have very different consequences on different targets (e.g., real-CPU architectures vs. WebAssembly or a hypothetical JVM target), so it might be nonintuitive what exactly it is guaranteeing.

So what exactly is the guarantee?  Well, I propose it should be roughly what I wrote [in this post](https://internals.rust-lang.org/t/annotations-for-zeroing-the-stack-of-sensitive-functions-which-deal-in-transient-secrets/11588/34): you define a mapping between the high-level Rust Abstract Machine state and the lower-level target state, and using `asm!` guarantees that the two states "sync up" to a limited extent.

In particular, `asm!` does not guarantee that no optimizations will take place.  Even physical CPUs do many things that could be seen as "optimizations", such as caching, speculative execution, and paging (the latter in coordination with the kernel).  And virtual machines can perform more aggressive optimizations, such as inlining.  (Note that by "virtual machines" I don't mean only e.g. WebAssembly implementations, but also emulators for real-CPU architectures.)

But those optimizations are not Rust's concern.  For a given target – say, `x86_64-unknown-linux-gnu`, `rustc` is defined as taking in Rust source files and outputting an executable. \*  The executable must follow certain requirements: in this case it must be in the ELF format, it must contain x86-64 instructions, and it should link against glibc, using the standard x86-64 ABI at the boundaries.  As a user, I can then run the executable in any environment that's capable of parsing that format, executing those instructions, and following that ABI.  I could run it under Linux, WSL, gVisor, or FreeBSD's Linux compatibility mode.  I could run it on a real CPU, or on qemu's JIT, or on a pure interpreter.  It's my choice.

If the Rust program contains an `asm!` statement, and that statement is reachable, then the executable must contain one or more copies of the specified instructions in its `.text` section (unless the user specifies a different section).  When the program reaches the asm statement, the native RIP (program counter) must be set to the address of one of those copies.  Any register inputs should have appropriate values; any memory that the assembly code is permitted to read (including anything transitively reachable through explicit inputs, as well as FFI-exposed global variables) must have its low-level state synced up with its Rust Abstract Machine state.

The implementation is not permitted to, say, bundle an x86 interpreter and run the assembly snippet through that.  It is not permitted to observe that `asm!("nop")` does nothing and eliminate it.  It *must* execute the instructions natively, exactly as specified.

Executing instructions is, itself, observable behavior.  If I use a debugger and set a breakpoint at each of the aforementioned copies of the assembly snippet, then when the program executes the `asm!` statement, the debugger must break at one of them.  I can then use the debugger to manually inspect the low-level state (i.e. register or memory contents); this must produce the right results for all low-level state that the assembly snippet was *permitted* to read, regardless of whether the snippet as written actually attempts to accesss it.  Similarly, I can modify any low-level state that the snippet was permitted to modify, and after continuing, the Rust Abstract Machine state must reflect those modifications.

Similarly, on a target whose native memory model does not support undefined bytes, executing an `asm!` block effectively `freeze`s all memory that the snippet is permitted to read and write.  This is simply because the implementation cannot distinguish 'the snippet left this byte alone' from 'the snippet wrote the same value to the byte that was already there'.  Just... don't be too quick to assume a target doesn't support undefined bytes.  For example, [in an example that has come up before](https://internals.rust-lang.org/t/what-the-hardware-does-is-not-what-your-program-does-uninitialized-memory/10561/17), Linux's `MADV_FREE` creates something like an undefined byte, and jemalloc uses it by default.

The requirement to sync high-level and low-level state does significantly constrain the optimizations the compiler can legally perform.  But it doesn't prevent all compiler optimizations, and the implementation guarantee is not specified in terms of "optimizations", since that would be hard to define precisely enough. 

Separately, the user's chosen CPU may perform its own optimizations, or, more generally, may not execute the instructions in constant time.  Indeed, most architectures do not provide any constant-time guarantees whatsoever.  On the other hand, if the user's CPU does somehow guarantee that certain assembly instructions will run in constant time, then an `asm!` block containing those instructions must run in constant time.  That's not a separate guarantee; it just follows from the guarantee that the specified instructions will run directly on the CPU.

\* rustc does support one "target" where the compilation process does not involve outputting an executable: miri.  The meaning of `asm!` on miri, of course, is not well-defined.

Reviewer 27: > The implementation is not permitted to, say, bundle an x86 interpreter and run the assembly snippet through that. It is not permitted to observe that asm!("nop") does nothing and eliminate it.  It must execute the instructions natively, exactly as specified.

I didn't say the compiler is allowed to *eliminate* the `nop` -- it is not, I agree with you here. But I don't see how from that you deduce that the compiler must not analyze the assembly. The example I brought was about the compiler realizing that the assembly "clobbers" less than what the programmer specified. To give a concrete example, I think the compiler is allowed to optimize `asm!("nop")` to `asm!("nop", flags(nomem))`. It is *not* allowed to remove the `asm!` block entirely.

The compiler has to ensure that the given bit of assembly runs natively, period. That doesn't mean it has to "sync up" any more than the assembly needs. The `asm!` macro specifies an *upper bound* of what the assembly needs synced, but if the compiler can figure out that less syncing is sufficient, then nothing I see in this spec prevents that.

My main concern here is the same as with `black_box`: it must not be possible to use `asm!` to inhibit optimizations and use that to "justify" UB. In other words, if a program has UB, then no amount of adding `asm!("nop")` to it can fix that UB. Are we in agreement about that at least, or not? Everything else, as far as I am concerned, is specifics.

Author: One way to think about it is: it is perfectly valid to, at runtime, patch out the nop and replace it with a function call. Even if the asm string is empty, it would be valid to set a breakpoint at its address and execute a signal handler when the breakpoint is hit.

I'm not saying that the compiler *cannot* analyze the asm, I'm saying that any analysis is useless because the compiler cannot assume that the instructions in the asm code are what is really being executed.

Author: With that said, such semantics only really apply to targets where the instruction stream is *observable*. This means that you can read the instruction bytes from within the program, get the current program counter position, interrupt the current thread with signals and inspect the register state at the interruption point, etc.

It does not apply to "soft" targets like WASM and asm.js where such details are not observable. On such targets the compiler has much more leeway for optimizing inline asm, for example by running `wasm-opt` after linking, which can make arbitrary changes to the generated instructions. This would be allowed because any such optimizations are not observable by the program itself when it is running. Incidentally this is one of the main reasons why I am hesitant to support `asm!` on such targets.

Reviewer 10: @Amanieu 
> With that said, such semantics only really apply to targets where the instruction stream is observable. This means that you can read the instruction bytes from within the program, get the current program counter position, interrupt the current thread with signals and inspect the register state at the interruption point, etc.

> It does not apply to "soft" targets like WASM and asm.js where such details are not observable. On such targets the compiler has much more leeway for optimizing inline asm, for example by running `wasm-opt` after linking, which can make arbitrary changes to the generated instructions. This would be allowed because any such optimizations are not observable by the program itself when it is running.

That's an excellent description, and one that I think should find itself into the RFC itself in some form. (I wouldn't call such targets "soft", but the concept seems sound.)

I think it makes complete sense to allow `asm!` on all targets that have any concept of assembly, including WebAssembly, with the above description explaining why WebAssembly may optimize the assembly if your workflow includes steps like `wasm-opt`.

Reviewer 16: > I didn't say the compiler is allowed to _eliminate_ the `nop` -- it is not, I agree with you here. But I don't see how from that you deduce that the compiler must not analyze the assembly.

Ah.  I thought you were implying that it could eliminate the nop, and so almost the entirety of my post is irrelevant.  Oh well. :p

Here is a better explanation of how I'd like to specify the part that *is* relevant:

The implementation is forbidden from assuming that the instructions in an `asm!` block will be executed 'faithfully' according to the architecture specification.  A CPU is allowed to do whatever it wants when it encounters those instructions, and the program must still operate correctly, as long as "whatever it wants" is limited to operations the asm block was *allowed* to perform.  "Allowed" is based on the specified inputs, outputt, clobbers, flags, etc., but not the asm string itself.

In practice, two reasons the CPU might be 'unfaithful' are if a debugger breakpoint is set at those instructions (an example I mentioned in my last post), or if the code has been patched (as Amanieu mentioned).

> > Similarly, on a target whose native memory model does not support undefined bytes, executing an asm! block effectively freezes all memory that the snippet is permitted to read and write.
> 
> This I do _not_ agree with. This is exactly the kind of thing that lets people use `asm!` blocks to cheat around UB. The best I could imagine guaranteeing here is that the `asm!` block freezes all memory _that it actually writes_.

It is indeed what lets people use `asm!` blocks to cheat around UB, and that's a good thing!  It's a powerful capability that can accomplish things that are otherwise impossible.  Admittedly, it's prone to misuse and misunderstanding, but that's a reason to write safer wrappers that expose useful subsets of functionality (...like what `black_box` should have been), not to ban it.

The main reason to ban it would be if it complicates the implementation, yet the compiler has no choice but to implement that capability.  In particular, even without `asm!`, FFI calls to unknown functions have almost the exact same semantics.  In the freezing example, since the function is unknown, the compiler has no foreknowledge of what memory the function is going to write.  It can try to determine it at runtime by comparing memory before and after the call, but that can only prove that the function *did* write to a given byte, not that it *didn't* (since it could have written the same value that was already there).  Thus, it must pessimistically assume that the function wrote data everywhere it was allowed to write.

Well, in theory the implementation could detect the write by setting a hardware watchpoint, but that's a pathological example that is not what anyone wants a compiler to do.

Reviewer 16: > It does not apply to "soft" targets like WASM and asm.js where such details are not observable. On such targets the compiler has much more leeway for optimizing inline asm, for example by running `wasm-opt` after linking, which can make arbitrary changes to the generated instructions. This would be allowed because any such optimizations are not observable by the program itself when it is running. Incidentally this is one of the main reasons why I am hesitant to support `asm!` on such targets.

Eh.  You can still attach a debugger to a wasm program, or patch it before loading it.  Both have legitimate use cases!  But even without the 'faithful execution' question, `wasm-opt` would still break the rules: it *can* eliminate `nop`s, for example.  I'm confused why it's necessary in the first place.  On every other architecture, we expect LLVM to generate optimal (fastest or smallest) code by itself, with no need for a post-processing step.  Is LLVM doing a bad job at codegen for WebAssembly, or is wasm-opt unnecessary?  Or are people using it in lieu of LTO?  The one real advantage of wasm-opt's approach is that its inputs are a stable format, so it can "LTO" against any library without needing to match compiler versions.  In exchange, though, it can't do as good a job at optimization, because WebAssembly is less information-rich than LLVM IR.

I think that ideally, we'd either commit to never having rustc automatically run `wasm-opt`, or teach `wasm-opt` to respect black-box semantics (creating some extension to mark code regions that are black boxes).  But I don't care very much about WebAssembly in general... so I'd be fine with giving `asm!` on WebAssembly weaker semantics than everywhere else, or just not supporting it at all.

Reviewer 21: > It might be worth clarifying that "accessible" here means "accessible according to Rust's aliasing rules". I.e., this is not allowed to perform any kind of access that unsafe Rust wouldn't also be permitted to do.

@RalfJung When reading or modifying inputs or outputs, which are necessarily visible to Rust, then I agree. 

But in general, I'm not sure I agree: an `asm!` block might read or write to global state not accessible to the Rust program in an unsynchronized way, causing a data-race. In Rust, the behavior of data-races is undefined, but on hardware, this isn't necessarily the case.

Reviewer 27: @Amanieu 

> One way to think about it is: it is perfectly valid to, at runtime, patch out the nop and replace it with a function call.

That... is extreme to say the least. Nothing in the RFC says that (as far as I can tell), and I am very concerned about the consequences of this.

> Even if the asm string is empty, it would be valid to set a breakpoint at its address and execute a signal handler when the breakpoint is hit.

Breakpoints can be set anywhere, not just in asm blocks, and the Rust spec just has nothing to say about what you will observe then. This is peeking into the mapping between the Abstract Machine and its implementation on actual hardware, and I don't think we want to specify this. So I don't think breakpoints or debuggers are a good argument.

@comex

> It is indeed what lets people use asm! blocks to cheat around UB, and that's a good thing!

In that case we are in visceral disagreement. To me this kind of treatment of UB looks extremely fragile and error prone and naive. When arguing for the UB-freedom of some code, you *don't* get to say that "yes this code has UB but it is okay because the compiler cannot see it", that's just not how UB works. *Even if* we end up modelling `asm!` as some kind of magic non-deterministic change to the Abstract Machine state that is constrained only by the asm block flags (something that IMO is rather extreme but seems to best fit your and @Amanieu's viewpoint), *even then* you still don't get to use that to "cheat around" UB. You still have to prove, for every execution of this non-deterministic execution that actually arises, that there is no UB according to the rules of the Abstract Machine.

Do you have a concrete example of the kind of "UB cheating" that you think ought to be legal?

> Admittedly, it's prone to misuse and misunderstanding, but that's a reason to write safer wrappers that expose useful subsets of functionality (...like what black_box should have been), not to ban it.

The thing that is "prone to misuse" here is not the UB cheating but the very spec itself -- IMO you are fundamentally breaking it by adding such a facility. *That* is why there was strong objection to this kind of "functionality" in the black_box discussion (not just by me), and *if* we decide to revise that it shouldn't happen hidden in the middle of a huge RFC that does all sorts of things.

black_box was spec'd the way it was for very good reasons. If you want to re-open that discussion, that's fine, but I do think it is *not* fine to do this "implicitly" as part of a rather large RFC that does not even mention this possibility.

The RFC should, for now, clearly state that UB cheating *remains disallowed*, and if we want to change that in the future we should have a discussion about that that is not burried in this particular RFC PR.

Reviewer 27: @gnzlbg 
> But in general, I'm not sure I agree: an asm! block might read or write to global state not accessible to the Rust program in an unsynchronized way, causing a data-race. In Rust, the behavior of data-races is undefined, but on hardware, this isn't necessarily the case.

Agreed, I was only referring to state that is observable inside the Abstract Machine.

## Operand type

Several types of operands are supported:

* `in(<reg>) <expr>`
  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
  - The allocated register will contain the value of `<expr>` at the start of the asm code.
  - The allocated register must contain the same value at the end of the asm code (except if a `lateout` is allocated to the same register).
* `out(<reg>) <expr>`
  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
  - The allocated register will contain an unknown value at the start of the asm code.
  - `<expr>` must be a (possibly uninitialized) place expression, to which the contents of the allocated register is written to at the end of the asm code.
  - An underscore (`_`) may be specified instead of an expression, which will cause the contents of the register to be discarded at the end of the asm code (effectively acting as a clobber).
* `lateout(<reg>) <expr>`
  - Identical to `out` except that the register allocator can reuse a register allocated to an `in`.

>   - You should only write to the register after all inputs are read, otherwise you may clobber an input.

Reviewer 27: Why is this "should", not "must"?

Also, is a violation of a "must" clause UB, or is there some other kind of failure mode?

  - The allocated register will contain the value of `<expr>` at the start of the asm code.
  - `<expr>` must be an initialized place expression, to which the contents of the allocated register is written to at the end of the asm code.
* `inout(<reg>) <in expr> => <out expr>`
  - Same as `inout` except that the initial value of the register is taken from the value of `<in expr>`.
  - `<out expr>` must be a (possibly uninitialized) place expression, to which the contents of the allocated register is written to at the end of the asm code.
  - An underscore (`_`) may be specified instead of an expression for `<out expr>`, which will cause the contents of the register to be discarded at the end of the asm code (effectively acting as a clobber).
  - `<in expr>` and `<out expr>` may have different types.
* `inlateout(<reg>) <expr>` / `inlateout(<reg>) <in expr> => <out expr>`
  - Identical to `inout` except that the register allocator can reuse a register allocated to an `in` (this can happen if the compiler knows the `in` has the same initial value as the `inlateout`).

>   - You should only write to the register after all inputs are read, otherwise you may clobber an input.

Reviewer 27: Another "should" that feels to me like violation might be UB?

- `nomem`: The `asm` blocks does not read or write to any memory. This allows the compiler to cache the values of modified global variables in registers across the `asm` block since it knows that they are not read or written to by the `asm`.
- `readonly`: The `asm` block does not write to any memory. This allows the compiler to cache the values of unmodified global variables in registers across the `asm` block since it knows that they are not written to by the `asm`.
- `preserves_flags`: The `asm` block does not modify the flags register (defined below). This allows the compiler to avoid recomputing the condition flags after the `asm` block.
- `noreturn`: The `asm` block never returns, and its return type is defined as `!` (never). Behavior is undefined if execution falls through past the end of the asm code.
- `nostack`: The `asm` block does not push data to the stack, or write to the stack red-zone (if supported by the target). If this flag is *not* used then the stack pointer is guaranteed to be suitably aligned (according to the target ABI) for a function call.

> 

Reviewer 21: Am I correct to assume that all `asm!` blocks can unwind if `-C panic=unwind` ? If so, I think it would be nice to have a `nounwind` flag here as well, such that if the `asm!` blocks unwinds, either the behavior is undefined, or the program terminates cc @rust-lang/wg-ffi-unwind .

Author: Since you are right that this is a potential footgun, we could add a `nounwind` flag and make it mandatory (it's a compile error if you don't specify it). At least until LLVM adds support for unwind from inline asm.

The downside is that now *every* asm! block needs to have this flag, which may be confusing to new users.

Reviewer 10: @Amanieu Given that the common case will be `nounwind`, let's instead consider adding an `unwind` flag if and when LLVM adds support for unwinding.

It would be incredibly noisy to need to annotate every asm block with such annotations, given that it's incredibly rare to want to unwind from asm.

Reviewer 21: > Since you are right that this is a potential footgun, we could add a nounwind flag and make it mandatory (it's a compile error if you don't specify it). At least until LLVM adds support for unwind from inline asm.

Sounds good to me.

* `lateout` operands with an `_` expression that are specified as an explicit register are converted to LLVM clobber constraints. For example, `lateout("r1") _` is mapped to `~{r1}` (cf. [llvm-clobber]).
* If the `nomem` flag is not set then `~{memory}` is added to the clobber list. (Although this is currently ignored by LLVM)
* If the `preserves_flags` flag is not set then the following are added to the clobber list:
  - (x86) `~{dirflag}~{flags}~{fpsr}`
  - (ARM/AArch64) `~{cc}`

For some operand types, we will automatically insert some modifiers into the template string.
* For `sym` and `imm` operands, we automatically insert the `c` modifier which removes target-specific modifiers from the value (e.g. `#` on ARM).
* On AArch64, we will warn if a value smaller than 64 bits is used without a modifier since this is likely a bug (it will produce `x*` instead of `w*`). Clang has this same warning.
* On ARM, we will automatically add the `P` or `q` LLVM modifier for `f64`, `v64` and `v128` passed into a `vreg`. This will cause those registers to be formatted as `d*` and `q*` respectively.

Additionally, the following attributes are added to the LLVM `asm` statement:

> * The `nounwind` attribute is always added: unwinding from an inline asm block is not allowed (and not supported by LLVM anyways).

Reviewer 21: This seems like a big footgun, e.g., if an `asm!` block calls a function that unwinds.

Reviewer 30: I would assume LLVM supports unwinding from inline assembly, otherwise why would it support using the `invoke` instruction as well as the `call` instruction to represent inline assembly in LLVM IR?

Reviewer 30: > I would assume LLVM supports unwinding from inline assembly, otherwise why would it support using the `invoke` instruction as well as the `call` instruction to represent inline assembly in LLVM IR?

Turns out that LLVM has `invoke asm` to help support unwinding from inline assembly, but [currently assumes inline assembly doesn't unwind](https://github.com/llvm/llvm-project/blob/master/llvm/test/Transforms/Inline/inline-invoke-with-asm-call.ll) and doesn't properly translate `invoke asm` in the backends.

Reviewer 10: > This seems like a big footgun, e.g., if an `asm!` block calls a function that unwinds the behavior is undefined. We could at least `abort` the program with a nice error message?

1) Aborting would require additional code around every asm block, for something that almost no asm block will *ever* do.

2) If LLVM doesn't support unwind from assembly, we can't do otherwise without adding such support to LLVM first.

3) We could potentially support an `unwind` attribute in the future, if LLVM adds such support.

Reviewer 21: > Aborting would require additional code around every asm block,

Can't we just use landing pads, just like we do for `#[unwind(abort)]`? That does increase code-size, and code size can impact execution time, but beyond this particular impact, the non-exceptional path should be fast.

> We could potentially support an unwind attribute in the future, if LLVM adds such support.

This would mean that some footguns are optin (nomem, nostack, pure, readonly, ...) while others are opt-out (unwind), and that feels inconsistent.

Reviewer 10: @gnzlbg
> > Aborting would require additional code around every asm block,
> 
> Can't we just use landing pads, just like we do for `#[unwind(abort)]`?

That would be one variant of the "additional code" I was referring to. And please no. Unwinding from assembly is incredibly rare, and the far far more common case of *not* unwinding from assembly should not incur overhead for it (whether by having to write "nounwind" on every asm block or by emitting a landing pad where we otherwise wouldn't have).

One of my use cases for inline assembly involves writing the smallest possible programs in Rust, using inline assembly to make system calls. (I hope to see crates for making syscalls become usable on stable soon after inline assembly becomes usable on stable.) Such programs would use `unwind=abort`, and should not contain any landing pads.

> > We could potentially support an unwind attribute in the future, if LLVM adds such support.
> 
> This would mean that some footguns are optin (nomem, nostack, pure, readonly, ...) while others are opt-out (unwind), and that feels inconsistent.

unwind isn't a footgun, here. Many assembly blocks will want to access memory, or the stack, or have side effects, or affect the flags; it's not at all unusual to need to do so, and having to pass a flag to allow doing so correctly is a footgun. The vast majority of inline assembly blocks don't unwind, and most don't even call a function; doing so is incredibly rare, and it doesn't seem at all unreasonable to require a flag to do so.

Author: > Can't we just use landing pads, just like we do for #[unwind(abort)]? That does increase code-size, and code size can impact execution time, but beyond this particular impact, the non-exceptional path should be fast.

That won't work because LLVM assumes that inline asm can't unwind. This means that it will optimize your landing pads away since they are "unreachable", or it may simply not emit the unwinding tables at all for the asm.

.p2align 2
.type foo_inline_asm, @function
foo_inline_asm:

// If necessary, save callee-saved registers to the stack here.
str x20, [sp, #-16]!

// Move the pointer to the argument out of the way since x0 is used.
mov x1, x0

// Load inputs values
ldr w2, [x1, #0]
ldr w0, [x1, #4]

<some asm code>

// Store output values
str w2, [x1, #0]
str w20, [x1, #8]

// If necessary, restore callee-saved registers here.
ldr x20, [sp], #16

ret

# Function suffix directives
.size foo_inline_asm, . - foo_inline_asm
```

And the following Rust code:

```rust
unsafe fn foo(mut a: i32, b: i32) -> (i32, i32)
{
    let c;
    {
        #[repr(C)]
        struct foo_inline_asm_args {
            a: i32,
            b: i32,
            c: i32,
        }
        extern "C" {
            fn foo_inline_asm(args: *mut foo_inline_asm_args);
        }
        let mut args = foo_inline_asm_args {
            a: a,
            b: b,
            c: mem::uninitialized(),
        };
        foo_inline_asm(&mut args);
        a = args.a;
        c = args.c;
    }
    (a, c)
}
```

[cranelift]: https://cranelift.readthedocs.io/
[cranelift-asm]: https://github.com/bytecodealliance/cranelift/issues/444

> 

Reviewer 21: ```suggestion
If `rustc` does not support inline assembly for a particular target and backend, trying to use the `asm!` macro on that particular backend and target fails to compile.
```

For example, consider we actually do this, and add support for using `asm!` with Cranelift and `x86_64`. What happens if Cranelift implements `aarch64` support, and somebody tries to compile a program for that target using that backend ? 

Should we forbid trying to do that? That is, should be forbid adding rustc support for new targets until `rustc` supports inline assembly for them? Or should we allow that as long as the programs being compiled do not use the `asm!` macro ? 

I think that blocking target/backend support on having inline assembly worked out imposes a too high barrier of entry for new targets. E.g., this RFC adds x86, arm, and risc-v support, but there are people using PowerPC and others on stable, which, according to this RFC, wouldn't initially gain any `asm!` macro support, yet we can't break that code.

Reviewer 7: My reading of that is simply that it is a compile error, not that the target can't be added at all.

Reviewer 21: Where did you read that ?

Reviewer 7: It says using `asm!` when building for a target without `asm` support is an error:
* So there must _be_ targets without `asm` support that otherwise can do builds.
* And building for them is fine as long as you don't use `asm`

Reviewer 21: Where does the RFC say that? (I've searched for that info a couple of times and can't find it, can you provide a link to the actual line?).

Reviewer 29: I think you are in violent agreement.  If I'm interpreting GitHub's interface properly @gnzlbg is suggesting to add the highlighted line since it isn't yet part of the RPC. Unfortunately, for other readers it looks like the comment is attached to an existing sentence from the proposal.

Reviewer 7: OH.

I thought the 2nd quote block was already in the RFC! Stupid GitHub interface.

Reviewer 21: Duh, I failed at this too. 

FWIW, @lokathor is much better at reading than me, and it is often the case that I miss something that's literally written "as is" in the text, and @lokathor just points me to it 😆 

So I actually was searching the whole RFC for this thinking "not again..." 😆

architectures a second-class citizen with respect to inline assembly.

## Provide intrinsics for each instruction

In discussions it is often postulated that providing intrinsics is a better solution to the
problems at hand.
However, particularly where precise timing, and full control over the number of generated
instructions is required intrinsics fall short.

Intrinsics are of course still useful and have their place for inserting specific instructions.
E.g. making sure a loop uses vector instructions, rather than relying on auto-vectorization.

However, inline assembly is specifically designed for cases where more control is required.
Also providing an intrinsic for every (potentially obscure) instruction that is needed
e.g. during early system boot in kernel code is unlikely to scale.

## Make the `asm!` macro return outputs

It has been suggested that the `asm!` macro could return its outputs like the LLVM statement does.
The benefit is that it is clearer to see that variables are being modified.
Particular in the case of initialization it becomes more obvious what is happening.
On the other hand by necessity this splits the direction and constraint specification from

> the variable name, which makes this syntax overall harder to read.

Reviewer 27: Instead of returning a tuple, this could return a (locally synthesized, anonymous) struct (I've recently seen a procedural macro do this). That would avoid having to split the variable name from the rest:
```rust
let res = asm!("mul {}", in(reg) a, in("eax") b, lateout("eax") lo, lateout("edx") ho);
res.hi as u64 << 32 + res.lo as u64
```

The main reason why this is so complicated is that GCC's inline assembly basically exports the raw internals of GCC's register allocator. This has resulted in many internal constraint codes and modifiers being widely used, despite them being completely undocumented.

## D & MSVC inline assembly

See the section [above][dsl].

# Unresolved questions
[unresolved-questions]: #unresolved-questions

None

# Future possibilities
[future-possibilities]: #future-possibilities

## Flag outputs

GCC supports a special type of output which allows an asm block to return a `bool` encoded in the condition flags register. This allows the compiler to branch directly on the condition flag instead of materializing the condition as a `bool`.

We can support this in the future with a special output operand type.

## `asm goto`

GCC supports passing C labels (the ones used with `goto`) to an inline asm block, with an indication that the asm code may jump directly to one of these labels instead of leaving the asm block normally.

> This could be supported by allowing code blocks to be specified as operand types. The following code will print `a` if the input value is `42`, or print `b` otherwise.

Reviewer 27: Alternatively, we could allow labels as arguments that can be jumped to, similar to https://github.com/rust-lang/rfcs/pull/2046?

Reviewer 10: @RalfJung That would make it substantially less structured. I'm hoping we can handle all use cases via what looks like a single multi-way branch similar to an if.

This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.

However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.

This can also be used with a general register class (e.g. `reg`) to obtain a scratch register for use inside the asm code:

```rust
// Multiply x by 6 using shifts and adds
let mut x = 4;
unsafe {
    asm!("
        mov {tmp}, {x}
        shl {tmp}, 1
        shl {x}, 2
        add {x}, {tmp}
    ", x = inout(reg) x, tmp = out(reg) _);
}
assert_eq!(x, 4 * 6);
```

## Register template modifiers

In some cases, fine control is needed over the way a register name is formatted when inserted into the template string. This is needed when an architecture's assembly language has several names for the same register, each typically being a "view" over a subset of the register (e.g. the low 32 bits of a 64-bit register).

```rust
let mut x: u16 = 0xab;

unsafe {

>     asm!("mov {0:h} {0:b}", inout(reg_abcd) x);

Reviewer 31: Shouldn't there be a comma here?
```suggestion
    asm!("mov {0:h}, {0:b}", inout(reg_abcd) x);
```

  - The allocated register will contain the value of `<expr>` at the start of the asm code.
  - The allocated register must contain the same value at the end of the asm code (except if a `lateout` is allocated to the same register).
* `out(<reg>) <expr>`
  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
  - The allocated register will contain an unknown value at the start of the asm code.
  - `<expr>` must be a (possibly uninitialized) place expression, to which the contents of the allocated register is written to at the end of the asm code.
  - An underscore (`_`) may be specified instead of an expression, which will cause the contents of the register to be discarded at the end of the asm code (effectively acting as a clobber).
* `lateout(<reg>) <expr>`
  - Identical to `out` except that the register allocator can reuse a register allocated to an `in`.
  - You should only write to the register after all inputs are read, otherwise you may clobber an input.

>   - `lateout` must be used instead of `out` if you are specifying the same explicit register as an `in`.

Reviewer 15: This has the same semantic meaning as having `inlateout` with the same register. I.e. `in("eax") <in_expr>, lateout("eax") <out_expr>` can be rewritten as `inlateout("eax") <in_expr> => <out_expr>`.

This should be at least noted here, but perhaps completely forbidden? There should be preferably one way to do a thing, and this rounds it out, mandating that a specific register can only participate in one directive. It also makes the code more readable, when you immediately see that the register is used for more than one purpose.

Author: This is a bit tricky. The reason why I decided to allow this is that in the case of `in(reg) <in_expr>, lateout(reg) <out_expr>` it is perfectly valid for the compiler to allocate the same register for both operands. In that spirit, it makes sense for `in("eax") <in_expr>, lateout("eax") <out_expr>` to work as well.

However you are right that this leaves us with two ways of doing the same thing, which could be confusing. I am not sure which option is the right way to go.

Reviewer 15: I think the difference, is that in the `reg` example, while you can't assume that they are not the same, you also can't assume that they are. Written out (contrived example): In
```
asm!("
mov {1}, {0}
add {1}, {0}
", in(reg) x, lateout(reg) y)
```
the `mov` is required; whereas in
```
asm!("
mov {1}, {0}
add {1}, {0}
", in("eax") x, lateout("eax") y)
```
It really isn't, but this isn't immediately clear. Refactoring,
```
asm!("
mov {0}, {0}
add {0}, {0}
", inlateout("eax") x => y)
```
make it immediately clear that it's redundant, and can therefore be removed.

Author: In your first example the `mov` is wrong regardless since it may clobber the input value if the same register is allocated to both operands.

Reviewer 30: > In your first example the `mov` is wrong regardless since it may clobber the input value if the same register is allocated to both operands.

not really, the first example does work when both `{0}` and `{1}` are the same register, as is demonstrated by the later example using `eax`. admittedly, this is somewhat of a special case where the input is overwritten by the same value.

Reviewer 15: Or we could change the example to
```
push {0}
pop {1}
```
if needed. The point is, that whereas `in(reg), lateout(reg)` and `inlateout(reg)` have different meanings (the second guarantees them to use the same register, the first doesn't) when using an explicit register this difference is eliminated (in both cases they are guaranteed to use the same, specified register). And since the second syntax is the one guaranteeing usage of the same register (with no other difference that I can tell), it makes sense for it to be used in the case of the specific register, where this also guaranteed.

Author: You make very good points and I have adjusted the RFC accordingly.

            a: a,
            b: b,
            c: mem::uninitialized(),
        };
        foo_inline_asm(&mut args);
        a = args.a;
        c = args.c;
    }
    (a, c)
}
```

## Rules for inline assembly
[rules]: #rules

- Any registers not specified as inputs will contain an undefined value on entry to the asm block.
- Any registers not specified as outputs must have the same value upon exiting the asm block as they had on entry.
  - This only applies to registers which can be specified as an input or output.
- Behavior is undefined if execution unwinds out of an asm block.
- Any memory reads/writes performed by the asm code follow the same rules as `volatile_read` and `volatile_write`.
  - Refer to the unsafe code guidelines for the exact rules.
  - If the `readonly` option is set, then only memory reads (with the same rules as `volatile_read`) are allowed.
  - If the `nomem` option is set then no reads or write to memory are allowed.
- Unless the `nostack` option is set, asm code is allowed to use stack space below the stack pointer.
  - On entry to the asm block the stack pointer is guaranteed to be suitably aligned (according to the target ABI) for a function call.

>   - You are responsible for making sure you don't overflow the stack (e.g. use stack probing to ensure you hit a guard page).

Reviewer 31: I think this should be carefully considered. It seems like this places an unnecessary burden on the user, especially if you don’t have control/knowledge of whatever stack protection Rust already decides to use. Also, since you cannot take advantage of whatever checks Rust already does, attempting to repeat such checks yourself would be a huge performance hit, but doing so could be necessary. Otherwise for example, what if you are using an embedded MCU with no stack protection or heap, create a large local array, and the array index overflows (somewhat feasible for 32-bit)? Do you need to check for overflow yourself at the start of each asm block, even though Rust may already check that there is enough space for what it knows about the stack frame (e.g. Rust local variables)?

One solution would be adding an option to reserve some amount of stack space. For example `stack(16)` (or with a better name) would reserve 16 bytes directly below the stack for you to use, i.e. at `esp - 0x10`.

The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.

This RFC only specifies how operands are substituted into the template string. Actual interpretation of the final asm string is left to the assembler.

However there is one restriction on the asm string: any assembler state (e.g. the current section which can be changed with `.section`) must be restored to its original value at the end of the asm string.

[rfc-2795]: https://github.com/rust-lang/rfcs/pull/2795

## Operand type

Several types of operands are supported:

* `in(<reg>) <expr>`
  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
  - The allocated register will contain the value of `<expr>` at the start of the asm code.
  - The allocated register must contain the same value at the end of the asm code (except if a `lateout` is allocated to the same register).
* `out(<reg>) <expr>`
  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.

>   - The allocated register will contain an unknown value at the start of the asm code.

Reviewer 27: I think elsewhere you call this "undefined value"? Is that the same as "unknown value"?
Both terminologies should be precisely defined. (I hope to avoid the situation in C where "indeterminate value" is full of ambiguities.) If I saw it correctly, you are only using "undefined value" in text that is written from the view of assembly, never on the level of the Abstract Machine, so I suspect the definition is "at the *beginning* of the execution of the asm block, non-deterministically pick any possible value this register can have in the target architecture" -- in particular, once the value is chosen, it is a "normal value", not something "unstable" that can change every time we look at it.

```
# Function prefix directives
.section ".text.foo_inline_asm"
.globl foo_inline_asm
.p2align 2
.type foo_inline_asm, @function
foo_inline_asm:

// If necessary, save callee-saved registers to the stack here.
str x20, [sp, #-16]!

// Move the pointer to the argument out of the way since x0 is used.
mov x1, x0

// Load inputs values
ldr w2, [x1, #0]
ldr w0, [x1, #4]

<some asm code>

// Store output values
str w2, [x1, #0]
str w20, [x1, #8]

// If necessary, restore callee-saved registers here.
ldr x20, [sp], #16

ret

# Function suffix directives
.size foo_inline_asm, . - foo_inline_asm
```

And the following Rust code:

```rust
unsafe fn foo(mut a: i32, b: i32) -> (i32, i32)
{
    let c;
    {
        #[repr(C)]
        struct foo_inline_asm_args {
            a: i32,
            b: i32,
            c: i32,
        }
        extern "C" {
            fn foo_inline_asm(args: *mut foo_inline_asm_args);
        }
        let mut args = foo_inline_asm_args {
            a: a,
            b: b,
            c: mem::uninitialized(),
        };
        foo_inline_asm(&mut args);
        a = args.a;
        c = args.c;
    }
    (a, c)
}
```

> ## Rules for inline assembly

Reviewer 27: Shouldn't this give some rules for lateout/inlateout?

[rules]: #rules

- Any registers not specified as inputs will contain an undefined value on entry to the asm block.

> - Any registers not specified as outputs must have the same value upon exiting the asm block as they had on entry.

Reviewer 27: Maybe be explicit about a violation of the "must" here being Undefined Behavior.

The compiler will choose an appropriate register to insert into
the template and will read the variable from there after the inline assembly finishes executing.

> Let us see another example that also uses an input:

Reviewer 32: ```suggestion
Let's see another example that also uses an input:
```

| AArch64 | `reg` | None | `i8`, `i16`, `i32`, `f32`, `i64`, `f64` |
| AArch64 | `vreg` | `fp` | `i8`, `i16`, `i32`, `f32`, `i64`, `f64`, <br> `i8x8`, `i16x4`, `i32x2`, `i64x1`, `f32x2`, `f64x1`, <br> `i8x16`, `i16x8`, `i32x4`, `i64x2`, `f32x4`, `f64x2` |
| ARM | `reg` | None | `i8`, `i16`, `i32`, `f32` |
| ARM | `sreg` | `vfp2` | `i32`, `f32` |
| ARM | `dreg` | `vfp2` | `i64`, `f64`, `i8x8`, `i16x4`, `i32x2`, `i64x1`, `f32x2` |
| ARM | `qreg` | `neon` | `i8x16`, `i16x8`, `i32x4`, `i64x2`, `f32x4` |
| RISC-V32 | `reg` | None | `i8`, `i16`, `i32`, `f32` |
| RISC-V64 | `reg` | None | `i8`, `i16`, `i32`, `f32`, `i64`, `f64` |
| RISC-V | `freg` | `f` | `f32` |
| RISC-V | `freg` | `d` | `f64` |

> Note: For the purposes of the above table pointers, function pointers and `isize`/`usize` are treated as the equivalent integer type (`i32` or `i64`).

If a value is of a smaller size than the register it is allocated in then the upper bits of that register will have an undefined value for inputs and will be ignored for outputs. The only exception is the `freg` register class on RISC-V where `f32` values are NaN-boxed in a `f64` as required by the RISC-V architecture.

> When separate input and output expressions are specified for an `inout` operand, both expressions must have the same type. The only exception is if both operands are pointers or integers, in which case they are only required to have the same size. This restriction exists register allocators in LLVM and GCC cannot handle tied operands with different types in some situations.

Reviewer 30: `This restriction exists register allocators` -> `This restriction exists because register allocators`

> Notes on allowed types:
> - Pointers and references are allowed where the equivalent integer type is allowed.

> > - `iLEN` refers to both sized and unsigned integer types. It also implicitly includes `isize` and `usize` where the length matches.

Reviewer 12: You fixed it to "sized and unsigned" instead of "signed and unsigned".

[llvm-clobber]: http://llvm.org/docs/LangRef.html#clobber-constraints
[issue-65452]: https://github.com/rust-lang/rust/issues/65452

## Rules for inline assembly
[rules]: #rules

> - Any registers not specified as inputs will contain an undefined value on entry to the asm block.

Reviewer 10: ...unless they are otherwise specified by the ABI. For instance, inline assembly on x86-64 Linux should be able to use fs to access thread-local storage (unless the developer has done something unusual to make that not work).

Reviewer 30: maybe it would be good to specify that the normally register-allocatable registers are undefined, but the other registers are defined to their standard ABI values, giving a exhaustive list somewhere, as that may be clearer.

For x86_64, the normally register-allocatable registers are:
* general purpose registers: rax, rcx, rdx, rbx, rbp, rsi, rdi, r8-r15
* SIMD registers: zmm0-31 and sub-registers
* mask registers: k0-k7
* rflags: df, of, zf, pf, sf, af
* mxcsr: pe, ue, oe, ze, de, ie
* st(0)-st(7)
* fsw
* any others I missed

Reviewer 30: Note that the list is different than the ABI, since that only applies at function boundaries. Inside functions, registers such as x86's direction flag that are defined by the ABI can be modified, but must be reset before the next function boundary.
